sympy/core/numbers.py

```diff
@@ -438,6 +439,142 @@ def ilcm(*args):
     return a
 
 
+def igcdLLL(s, m1=2, n1=3, optimize=False):
+    """return g, B where g is the gcd of the integers in s and the
+    LLL-reduced basis and small multipliers such that those
+    multiples of the corresponding values of s is g.
+
+    If there are three values in s then the smallest multipliers
+    will be returned unless optimize is not True.
+
+    This is an iterative procedure whose quality in returning
+    small multipliers is related to alpha = m1/n1 which must
+    be in the range [3/8, 1]. When 1, the algorithm will iterate
+    longer but is not guaranteed to return the smallest multipliers
+    except for the case of 3 values in s.
+
+    Examples
+    ========
+
+    >>> from sympy.core.numbers import igcdLLL
+    >>> from sympy import Add, Matrix, symbols
+
+    >>> s = 2, 4, 6
+    >>> x = symbols('x:%s' % len(s))
+    >>> eq = Add(*[si*xi for si, xi in zip(s, x)]); eq
+    2*x0 + 4*x1 + 6*x2
+
+    >>> g, B = igcdLLL(s)
+    >>> assert Add(*[si*mi for si, mi in zip(s, B[-1])]) == g
+
+    >>> p = symbols('p:%s' % (len(B) - 1))
+    >>> psol = (Matrix(B[:-1]).T*Matrix(p)).T.tolist()[0]
+    >>> dict(zip(x, psol))
+    {x0: p0 + 2*p1, x1: p0 - p1, x2: -p0}
+    >>> eq.subs(_)
+    0
+    """
+    if not iterable(s):
+        raise ValueError('s should be iterable')
+    if len(s) < 2:
+        raise ValueError('expecting 2 or more values')
+    if not all(type(i) is int for i in s):
+        raise ValueError('expecting literal ints for s')
+    ok = (3*n1 <= 8*m1) and (n1 > 0) and (m1 > 0) and (m1 <= n1)
+    if not ok:
+        raise ValueError('require 3/8 <= alpha <= 1 but alpha = %s/%s' % (m1, n1))
+    if optimize and len(s) > 3:
+        raise ValueError('optimization for more than 3 values')
+
+    def reduce1(k, i):
+        if a[i] != 0:
+            q = int(round(a[k] / a[i]))
+        else:
+            if 2 * abs(la[k][i]) > D[i]:
+                q = int(round(la[k][i] / D[i]))
```

Comment:
There are a few places here that do `int / int -> float` division. Is it possible to write this without floating point division? This should presumably fail in some cases for very large integers (`> 2^53`).

Reply:
I was thinking the same and had each of those lines tested in the version I tried to translate (without success) from Rosetta Code.

Another issue is the `round`. Python rounds away from zero. I'm not sure that this will make a difference. The HMM paper uses `\ceiling x \floor` and the translations on the (Rosetta Code page]() uses the round-away-even-and-odd-the-same sort of `round`, too. So we might need a custom `round` to handle unlimited integer size:
```python
def iround(num, den):
  # round away from 0 on tie
  if num < 0:
    return -iround(-num, den)
  w, r =divmod(num, den)
  r2 = 2*r
  if r2 < den:
    return w
  return w + 1
```

Reply:
Changes to that effect were made.

Here is where there is division which is not guaranteed to be integer:
```
        for i in range(k + 1, m):
            t = la[i][j]*D[k] - la[i][k]*la[k][j]
            #la[i][j] = (la[i][j]*la[k][j] + la[i][k]*D[j - 1])/D[j]
            la[i][j] = round_out((la[i][j]*la[k][j] + la[i][k]*D[j - 1]),D[j])
            #assert la[i][j] == int(la[i][j]), la[i][j] # [-9203, -9333, 3922] -1927141.000024883
            #la[i][k] = t/D[j]
            la[i][k] = round_out(t,D[j])
            #assert la[i][k] == int(la[i][k]), la[i][k] # [9221, -7615, -7571],-70482653.00000001
        #D[j] = (D[k - 2]*D[k] + la[k][j]**2)/D[j]
        D[j] = round_out((D[k - 2]*D[k] + la[k][j]**2),D[j])
        #assert D[j] == int(D[j]), D[j] # [9997, 9079, -5001], 182368249.99999997
```
I used the round function on them but I am not sure of the implications of this in terms of the algorithm. All I have confirmed so far is that after running for a while with 3-tuples of values in the range of [-10**4, 10**4] it doesn't hang.

Reply:
This looks interesting. You mentioned a paper. Do you have the reference?

I'm wondering about the division with rounding. I'm just surprised that it's not ordinary exact division, i.e. `q = a[k] // a[i]`.

Reply:
@skieffer , refs are in the OP.  They will be added to the commit later. I still need to link this up to the `diophantine` module. There are many more applications for this as discussed on the Rosetta Code link (given in the OP).

Reply:
@skieffer , the expectation is that division is exact or is explicitly floor-related. I will update the code. The reason I was getting floats is because ints-as-floats lost precision at some point and then the assertion that `int(afloat) == afloat` began to fail.